code/game/objects/structures/universal_generators/cellular_automata.dm

```diff
@@ -0,0 +1,214 @@
+	var/turf/ca_wall
+	var/turf/ca_floor
+	var/mapgrid_width = 40
+	var/mapgrid_height = 21
+	var/percent_area_walls = 45
+	var/iterations = 5
+	var/watch
+	var/mapgrid_scale = 1
+	var/list/list_of_turfs = list()
+	var/list/mapgrid
+	var/could_not_place = 0
+
+/obj/structure/radial_gen/cellular_automata/Destroy()
+	..()
+	mapgrid.Cut()
+	list_of_turfs.Cut()
+
+/obj/structure/radial_gen/cellular_automata/deploy_generator(var/turf/bottomleft)
+	init_mapgrid() // first generates a random map
+	for(var/i = 1 to iterations)
+		make_caverns(i)
+	could_not_place = check_mapgrid(bottomleft)
+	if(!could_not_place)
+		apply_mapgrid_to_turfs(bottomleft)
+
+/obj/structure/radial_gen/cellular_automata/proc/init_mapgrid()
+	if(!mapgrid)
+		mapgrid = new/list(mapgrid_width,mapgrid_height)
+	else if(could_not_place)
+		could_not_place = 0
+		return
+	for(var/row = 1 to mapgrid_height)
+		for(var/column = 1 to mapgrid_width)
+			if(column == 1)
+				mapgrid[column][row] = CA_PERMAWALL
+			else if(row == 1)
+				mapgrid[column][row] = CA_PERMAWALL
+			else if(column == mapgrid_width)
+				mapgrid[column][row] = CA_PERMAWALL
+			else if(row == mapgrid_height)
+				mapgrid[column][row] = CA_PERMAWALL
+			else // otherwise, fill the walls a random percent of the time
+				var/random_number = rand(1,100)
+				if(random_number < percent_area_walls)
+					mapgrid[column][row] = CA_WALL
+				else
+					mapgrid[column][row] = CA_FLOOR
+
+/obj/structure/radial_gen/cellular_automata/proc/make_caverns(var/iteration)
+	for(var/row = 1 to mapgrid_height)
+		for(var/column = 1 to mapgrid_width)
+			mapgrid[column][row] = place_wall_logic(column,row,iteration)
+
+/obj/structure/radial_gen/cellular_automata/proc/place_wall_logic(var/column,var/row,var/iteration)
+	var/num_walls = get_adjacent_walls(column,row)
+	switch(mapgrid[column][row])
+		if(CA_PERMAWALL)
+			return CA_PERMAWALL
+		if(CA_WALL)
+			if(num_walls >= 4)
+				return CA_WALL
+		if(CA_FLOOR)
+			if(num_walls >= 5)
+				return CA_WALL
+	return CA_FLOOR
+
+/obj/structure/radial_gen/cellular_automata/proc/get_adjacent_walls(var/column,var/row,var/n = 1)
+
+	if(mapgrid[column][row] == CA_PERMAWALL)
+		return
+
+	var/wallcounter = 0
+
+	for(var/iy = (row-n) to (row+n))
+		for(var/ix = (column-n) to (column+n))
+			if(ix > 0 && ix < mapgrid_width)
+				if(!(ix == column && iy == row) && iy > 0 && iy < mapgrid_height)
+					if(mapgrid[ix][iy])
+						wallcounter++
+	return wallcounter
+
+
+/obj/structure/radial_gen/cellular_automata/proc/check_mapgrid(var/turf/bottomleft)
+	for(var/row = 1 to mapgrid_height)
+		for(var/column = 1 to mapgrid_width)
+			if(!(mapgrid[column][row])) // it's a floor
+				for(var/scale_row = 1 to mapgrid_scale)
+					for(var/scale_column = 1 to mapgrid_scale)
+						var/turf/T = locate((bottomleft.x-1)+(column * mapgrid_scale)+scale_column,(bottomleft.y-1)+(row * mapgrid_scale) + scale_row,bottomleft.z)
+						if(T && istype(T,ca_wall))
+							continue
+						else
+							return TRUE
```

Comment:
Pretty certain you can negate the `if` clause and skip the `else` completely : 
```
if (x)
     continue
else
     return TRUE
```
is equivalent to
```
if (!x)
     return TRUE
```
as far as I'm aware